### ISSUE TITLE
clangd + semanticHighlight causes major lag and slow down

### ISSUE THREAD

**Issue by S4deghN** [yegappan/lsp#491]
takes seconds to delete a line or undo it same for loading buffers.
just disabling `semanticHighlight` make everything snappy.
any idea what's the issue?

this is my config for reference:
```
var lspOpts = {
    semanticHighlight: true,
    aleSupport: false,
    autoComplete: true,
    autoHighlight: false,
    autoHighlightDiags: true,
    autoPopulateDiags: false,
    completionMatcher: 'icase',
    completionMatcherValue: 1,
    echoSignature: false,
    hideDisabledCodeActions: false,
    highlightDiagInline: false,
    hoverInPreview: false,
    ignoreMissingServer: true,
    keepFocusInDiags: false,
    keepFocusInReferences: true,
    completionTextEdit: true,
    diagVirtualTextAlign: 'above',
    diagVirtualTextWrap: 'default',
    noNewlineInCompletion: false,
    omniComplete: false,
    outlineOnRight: false,
    outlineWinSize: 20,
    showDiagInBalloon: true,
    showDiagInPopup: true,
    showDiagOnStatusLine: false,
    showDiagWithSign: true,
    showDiagWithVirtualText: false,
    showInlayHints: false,
    showSignature: true,
    snippetSupport: false,
    ultisnipsSupport: false,
    useBufferCompletion: false,
    usePopupInCodeAction: false,
    useQuickfixForLocations: true,
    vsnipSupport: false,
    bufferCompletionTimeout: 0,
    customCompletionKinds: false,
    completionKinds: {},
    filterCompletionDuplicates: false,
}

var lspServers = [
    {
        name: 'clangd',
        filetype: ['c', 'cpp'],
        path: 'clangd',
        args: [
            '--background-index',
            "--malloc-trim",
            "--enable-config",
            "--all-scopes-completion=true",
            "--completion-style=detailed",
            "--function-arg-placeholders",
            "--header-insertion=iwyu",
            "--header-insertion-decorators",
        ]
    },
]

```

**Comment by S4deghN** [comment]
I wanted to add that syntax highlighting on small projects doesn't introduce any discernible lag.
And that I also tested the same project with a minimal neovim with native lsp config to see if it's a clangd issue but neovim didn't have any problem with syntax highlighting.

**Comment by MisterSnuvi** [comment]
I tried debugging this as I was experiencing that myself. It looks to me that on many changes, e.g. when "dd" is spammed, the synchronous server calls cannot keep up and block for a longer period. I fixed it on my side by making the call at [lspserver.vim#L564](https://github.com/yegappan/lsp/blob/main/autoload/lsp/lspserver.vim#L564) asynchronous and calling "UpdateTokens" in the callback. I am not sure if this is a good solution as I am not aware of the reason for a synchronous call in the first place (maybe just to prevent race conditions?).

**Comment by eithrial** [comment]
@MisterSnuvi could you provide they way you fixed this? I tried making it asynchronous, but results using my solution were mixed, so I still keep semantic highlighting off.

**Comment by MisterSnuvi** [comment]
@eithrial I changed the lines from [lspserver.vim#564](https://github.com/yegappan/lsp/blob/main/autoload/lsp/lspserver.vim#L564) to 568 to 
```
564   │   lspserver.rpc_a(method, params, (_, reply) => {
565   │     if reply->empty()
566   │       return
567   │     endif
568   │     semantichighlight.UpdateTokens(lspserver, bnr, reply)
569   │   })
```
It's not perfectly stable with this but the lags are mostly gone.

**Comment by S4deghN** [comment]
> @eithrial I changed the lines from [lspserver.vim#564](https://github.com/yegappan/lsp/blob/main/autoload/lsp/lspserver.vim#L564) to 568 to
> 
> ```
> 564   │   lspserver.rpc_a(method, params, (_, reply) => {
> 565   │     if reply->empty()
> 566   │       return
> 567   │     endif
> 568   │     semantichighlight.UpdateTokens(lspserver, bnr, reply)
> 569   │   })
> ```
> 
> It's not perfectly stable with this but the lags are mostly gone.

I want to confirm that this diminishes the lag for me too, but it's not all gone.
I basically gave up using the semanticHighlight since I encountered this problem but besides that the plugin works really well.

**Comment by briancarbone** [comment]
It does seem as though undoing nearly always causes Vim to hang for 1-3 full seconds. I wasn’t sure whether it was LSP, but this is certainly evidence…

**Comment by yegappan** [comment]
> [@yegappan](https://github.com/yegappan) Is it possible you could make semantic highlighting asynchronous? I think the synchronous nature causes Vim to hang when undoing or making larger changes.

Yes.  Let me look into making the semantic highlighting asynchronous.

**Comment by Konfekt** [comment]
At some point it also starts throwing on every change (in a Svelte file)

```vim
Error detected while processing TextChanged Autocommands for "<buffer=136>"..function <SNR>437_LspUpdateSemanticHighlight[6]..<SNR>417_SemanticHighlightUpdate[32]..lsp#semantichighlight#UpdateTokens:
line   35:
E964: Invalid column number: 30
```

Maybe these can be suppressed?
